torch/quantization/observer.py

```diff
@@ -1,50 +1,62 @@
-    that computes the quantization parameters given the collected statistics.
-    TODO: Maybe add an abstract Observer class that enforces these rules?
+import torch
+import torch.nn as nn
+
+
+class ObserverBase(ABC, nn.Module):
+    r"""Observer base Module
+    Any concrete observer implementation should derive from this class.
+
+    Concrete observers should follow the same API. In forward, they will update
+    the statistics of the observed Tensor. And they should provide a
+    `calculate_qparams` function that computes the quantization parameters given
+    the collected statistics.
     """
+
     def __init__(self, dtype=torch.quint8, qscheme=torch.per_tensor_affine):
-        super(Observer, self).__init__()
+        super(ObserverBase, self).__init__()
         self.dtype = dtype
         self.qscheme = qscheme
-        assert self.qscheme in (torch.per_tensor_affine, torch.per_tensor_symmetric), \
-            'Default Observer only works for per_tensor_affine and \
-                per_tensor_symmetric quantization scheme'
-        assert self.dtype in (torch.qint8, torch.quint8), \
-            'Default Observer only works for qint8 and quint data type'
-        self.min_val = None
-        self.max_val = None
+        assert self.qscheme in (
+            torch.per_tensor_affine,
+            torch.per_tensor_symmetric,
+        ), "Default Observer only works for per_tensor_affine and \
+                per_tensor_symmetric quantization scheme"
+        assert self.dtype in (
+            torch.qint8,
+            torch.quint8,
+        ), "Default Observer only works for qint8 and quint data type"
 
+    @abstractmethod
     def forward(self, x):
-        if self.min_val is None or self.max_val is None:
-            self.min_val = torch.min(x)
-            self.max_val = torch.max(x)
-        else:
-            self.min_val = torch.min(torch.min(x), self.min_val)
-            self.max_val = torch.max(torch.max(x), self.max_val)
+        pass
+
+    @abstractmethod
+    def calculate_qparams(self, **kwargs):
+        pass
+
+    def _calculate_qparams(self, min_val, max_val):
+        """
+        Given min and max values, this function calculates quantization parameters
+        """
+        assert min_val <= max_val, "min {} should be less than max {}".format(
+            min_val, max_val
+        )
 
-    def calculate_qparams(self):
         if self.dtype == torch.qint8:
             qmin, qmax = -128, 127
         else:
             qmin, qmax = 0, 255
-        n_levels = 255.0
-        if self.max_val is None or self.min_val is None:
-            raise Exception('must run observer before calling calculate_qparams!')
-        max_val, min_val = self.max_val.item(), self.min_val.item()
+        n_levels = qmax - qmin
+
+        # extend min/max values to include 0 to meet the requirement that 0 is
+        # exactly repsentable
+        min_val = min(min_val, 0.0)
```

Comment:
Dont you need torch.min here?

Reply:
I saw torch.min is to get the min value of a tensor, here is to get the min of two numbers, do we need torch.min in this case?